src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,230 @@
+    >>;
+
+  virtual write_iertr::future<> write(std::list<LogicalCachedExtentRef>& extent) = 0;
+  virtual ~ExtentOolWriter() {}
+};
+
+class SegmentedOolWriter : public ExtentOolWriter,
+                          public boost::intrusive_ref_counter<
+  SegmentedOolWriter, boost::thread_unsafe_counter>{
+public:
+  using roll_segment_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using alloc_extent_ertr = roll_segment_ertr;
+  using init_segment_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+
+  SegmentedOolWriter(SegmentProvider& sp, SegmentManager& sm)
+    : segment_provider(sp), segment_manager(sm) {}
+  write_iertr::future<> write(std::list<LogicalCachedExtentRef>& extent) final;
+
+private:
+  bool _needs_roll(segment_off_t length) const;
+  roll_segment_ertr::future<> roll_segment();
+  init_segment_ertr::future<> init_segment(Segment& segment);
+
+  using extents_to_write_t = std::vector<LogicalCachedExtentRef>;
+  void add_extent_to_write(
+    ool_record_t&,
+    LogicalCachedExtentRef& extent);
+
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  SegmentRef current_segment;
+  std::vector<SegmentRef> open_segments;
+  segment_off_t allocated_to = 0;
+};
+
+using SegmentedOolWriterRef = std::unique_ptr<SegmentedOolWriter>;
+
+class scan_valid_records_cursor;
+
+template <typename HintT = empty_hint_t>
+class ExtentAllocator {
+public:
+  using scan_device_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  virtual CachedExtentRef alloc_ool_extent(
+    Transaction& t,
+    extent_types_t type,
+    const HintT hint,
+    segment_off_t length) = 0;
+
+  virtual ~ExtentAllocator() {};
+};
+
+template <typename HintT = empty_hint_t>
+using ExtentAllocatorRef = std::unique_ptr<ExtentAllocator<HintT>>;
+
+template <typename IndexT, typename HintT = empty_hint_t>
+class SegmentedAllocator : public ExtentAllocator<HintT> {
+public:
+  using calc_target_func_t = typename std::function<IndexT (HintT)>;
+  using alloc_extent_ertr = SegmentedOolWriter::alloc_extent_ertr;
+  SegmentedAllocator(
+    SegmentProvider& sp,
+    SegmentManager& sm,
+    Cache& cache,
+    calc_target_func_t&& calc_target_func)
+    : segment_provider(sp),
+      segment_manager(sm),
+      cache(cache),
+      calc_target_func(std::move(calc_target_func))
+  {}
+
+  CachedExtentRef alloc_ool_extent(
+    Transaction& t,
+    extent_types_t type,
+    const HintT hint,
+    segment_off_t length) final {
+    auto index = calc_target_func(hint);
+    auto iter = writers.find(index);
+    if (iter == writers.end()) {
+      iter = writers.emplace(
+        index,
+        std::make_unique<SegmentedOolWriter>(
+          segment_provider,
+          segment_manager)).first;
+    }
+    auto& writer = iter->second;
+
+    auto nextent = cache.alloc_new_extent_by_type(
+      t, type, length, paddr_t{ZERO_SEG_ID, fake_paddr_off});
```

Comment:
> IIUC the logical dependency needs to be reversed so that `Cache` depends on `ExtentPlacementManager` to allocate new extents. Then `ExtentPlacementManager` can decide whether to allocate ool or inline extents according to the hint.

Currently, extents that are writing to journal are allocated by the transaction manager, not directly by cache, and the Cache doesn't depend on the transaction manager. With this change, the situation would be extents for journal are still allocated by the transaction manager, while ool extents are allocated by the extent placement manager. Cache still doesn't rely on any other structures. I think this is more consistent, and in the initial design doc we also discussed this.

Reply:
No, the "allocation time" I was mentioning is when the concrete extent is constructed in memory by `Cache` with relative paddr, before it is allocated physically during transaction submission (IIUC your "allocation time" concept):
https://github.com/ceph/ceph/blob/fa816c05868d311d98e4f6c95d8d982c74f12fac/src/crimson/os/seastore/cache.cc#L671-L675

I think this is where ExtentPlacementManager can decide whether the extent is inlined or ool according to user hint.

Reply:
I know. `Cache::alloc_new_extent_by_type` is called by LBAManager for rewrites, and I changed it to be called by ExtentPlacementManager, also for rewrites. I think this is more consistent. Am I right? Before this pr, new extents are allocated by the transaction manager, rewritten extents are allocated by lba_manager, all through `Cache::alloc_new_extent`; now, new extents are allocated by the transaction manager, rewritten extents are allocated by the extent placement manager, still all through `Cache::alloc_new_extent`. And if we decide to let the extent placement manager to decide whether an extent is inline in the future, all extents will be allocated by the extent placement manager, again through `Cache::alloc_new_extent`. I think this is a more consistent and easy-to-maintain coding pattern. Right?

Reply:
I think we can go back to this discussion after https://github.com/ceph/ceph/pull/41803#discussion_r674463545